code/modules/paperwork/paper.dm

```diff
@@ -153,13 +153,22 @@
 	else
 		. += "<span class='notice'>You have to go closer if you want to read it.</span>"
 
-/obj/item/weapon/paper/proc/show_content(var/mob/user, var/forceshow=0)
-	if(!(istype(user, /mob/living/carbon/human) || istype(user, /mob/observer/dead) || istype(user, /mob/living/silicon)) && !forceshow)
-		user << browse("<HTML><HEAD><TITLE>[name]</TITLE></HEAD><BODY>[stars(info)][stamps]</BODY></HTML>", "window=[name]")
-		onclose(user, "[name]")
-	else
-		user << browse("<HTML><HEAD><TITLE>[name]</TITLE></HEAD><BODY>[info][stamps]</BODY></HTML>", "window=[name]")
-		onclose(user, "[name]")
+/obj/item/weapon/paper/proc/show_content(mob/user, forceshow)
+	var/datum/browser/paper_win = new(user, name, null, 450, 500, null, TRUE)
+	paper_win.set_content(get_content(user, can_read(user, forceshow)))
+	paper_win.add_stylesheet("paper_languages", 'html/browser/paper_languages.css')
+	paper_win.open()
+
+/obj/item/weapon/paper/proc/can_read(var/mob/user, var/forceshow = FALSE)
+	var/can_read = ((istype(user, /mob/living/carbon/human) || istype(user, /mob/observer/dead) || istype(user, /mob/living/silicon)) || forceshow)
+	if(!forceshow && istype(user,/mob/living/silicon/ai))
+		var/mob/living/silicon/ai/AI
+		can_read = get_dist(src, AI.camera) < 2
+	return can_read
+
+/obj/item/weapon/paper/proc/get_content(var/mob/user, var/can_read = TRUE)
+	return "<head><title>[capitalize(name)]</title><style>body {background-color: [color];}</style></head><body>[can_read ? parse_languages(user, info) : stars(info)][stamps]</body>"
```

Comment:
In the case where the user can't read it at all, do we really have to process the whole string? Can we not just prefabricate some generic "You cannot read this" response?